File: workers/_shared/acfworker/worker.py
from __future__ import annotations

import json
import logging
import signal
import sys
import threading
import time
from dataclasses import dataclass
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from typing import Any, Callable

import redis

from .api import BackendApi
from .config import WorkerConfig, load_config
from .queue import JobQueue, JobRecord, WORKER_SET, worker_key
from .storage import Storage

POLL_INTERVAL_SEC = 1.0
MAINTENANCE_INTERVAL_SEC = 15.0
HEARTBEAT_INTERVAL_SEC = 10.0
HEARTBEAT_TTL_SEC = 30


class WaitingForGpu(Exception):
    def __init__(self, message: str, retry_in_ms: int) -> None:
        super().__init__(message)
        self.retry_in_ms = retry_in_ms


class PermanentJobError(Exception):
    pass


class JsonFormatter(logging.Formatter):
    def __init__(self, service: str, worker_id: str) -> None:
        super().__init__()
        self.service = service
        self.worker_id = worker_id

    def format(self, record: logging.LogRecord) -> str:
        payload: dict[str, Any] = {
            "level": record.levelname.lower(),
            "time": self.formatTime(record, "%Y-%m-%dT%H:%M:%S%z"),
            "service": self.service,
            "workerId": self.worker_id,
            "msg": record.getMessage(),
        }
        extra = getattr(record, "extra", None)
        if isinstance(extra, dict):
            payload.update(extra)
        if record.exc_info:
            payload["error"] = self.formatException(record.exc_info)
        return json.dumps(payload, ensure_ascii=False)


def build_logger(config: WorkerConfig) -> logging.Logger:
    logger = logging.getLogger(config.worker_name)
    logger.setLevel(getattr(logging, config.log_level.upper(), logging.INFO))
    handler = logging.StreamHandler(sys.stdout)
    handler.setFormatter(JsonFormatter(config.worker_name, config.worker_id))
    logger.handlers = [handler]
    logger.propagate = False
    return logger


@dataclass
class JobContext:
    job: JobRecord
    data: dict[str, Any]
    config: WorkerConfig
    api: BackendApi
    storage: Storage
    logger: logging.Logger
    cancelled: threading.Event
    _report: Callable[[float, str | None], None]

    def report_progress(self, progress: float, message: str | None = None) -> None:
        self._report(progress, message)


class HealthServer:
    def __init__(self, port: int, probe: Callable[[], tuple[bool, dict[str, Any]]]) -> None:
        self.probe = probe

        outer = self

        class Handler(BaseHTTPRequestHandler):
            def do_GET(self) -> None:  # noqa: N802
                if self.path not in ("/", "/health", "/healthz"):
                    self.send_response(404)
                    self.end_headers()
                    return
                try:
                    ok, detail = outer.probe()
                except Exception as exc:  # noqa: BLE001
                    ok, detail = False, {"error": str(exc)}
                body = json.dumps({"status": "ok" if ok else "degraded", **detail}).encode()
                self.send_response(200 if ok else 503)
                self.send_header("content-type", "application/json")
                self.send_header("content-length", str(len(body)))
                self.end_headers()
                self.wfile.write(body)

            def log_message(self, *_args: Any) -> None:
                return

        self.server = ThreadingHTTPServer(("0.0.0.0", port), Handler)
        self.thread = threading.Thread(target=self.server.serve_forever, daemon=True)

    def start(self) -> None:
        self.thread.start()

    def stop(self) -> None:
        self.server.shutdown()


def run_worker(
    name: str,
    queue: str,
    handler: Callable[[JobContext], dict[str, Any]],
    on_start: Callable[[WorkerConfig, logging.Logger], dict[str, Any]] | None = None,
    default_concurrency: int = 1,
) -> None:
    config = load_config(name, default_concurrency)
    logger = build_logger(config)

    client = redis.from_url(config.redis_url, decode_responses=True, socket_keepalive=True)
    job_queue = JobQueue(client)
    api = BackendApi(config.backend_url, config.internal_api_key, config.worker_id)
    storage = Storage(config.data_dir, config.storage_mode, api)

    capabilities: dict[str, Any] = {}
    startup_error: str | None = None
    if on_start is not None:
        try:
            capabilities = on_start(config, logger) or {}
        except Exception as exc:  # noqa: BLE001
            startup_error = str(exc)
            logger.error("Initialisierung fehlgeschlagen, Worker laeuft eingeschraenkt weiter", exc_info=True)

    state = {
        "status": "idle",
        "reason": startup_error,
        "current_job": None,
        "processed": 0,
        "failed": 0,
        "started_at": int(time.time() * 1000),
    }
    stop_event = threading.Event()

    def publish_heartbeat() -> None:
        payload = {
            "id": config.worker_id,
            "name": config.worker_name,
            "queue": queue,
            "version": config.version,
            "host": config.host,
            "startedAt": state["started_at"],
            "lastSeen": int(time.time() * 1000),
            "status": state["status"],
            "currentJobId": state["current_job"],
            "concurrency": config.concurrency,
            "capabilities": capabilities,
            "reason": state["reason"],
        }
        pipe = client.pipeline()
        pipe.set(worker_key(config.worker_id), json.dumps(payload), ex=HEARTBEAT_TTL_SEC)
        pipe.sadd(WORKER_SET, config.worker_id)
        pipe.execute()

    def heartbeat_loop() -> None:
        while not stop_event.is_set():
            try:
                publish_heartbeat()
            except Exception:  # noqa: BLE001
                pass
            stop_event.wait(HEARTBEAT_INTERVAL_SEC)

    def maintenance_loop() -> None:
        while not stop_event.is_set():
            try:
                job_queue.promote_delayed(queue)
                recovered = job_queue.reap_stalled(queue)
                if recovered:
                    logger.warning(
                        "Jobs abgestuerzter Worker wiederhergestellt",
                        extra={"extra": {"jobIds": recovered}},
                    )
            except Exception:  # noqa: BLE001
                pass
            stop_event.wait(MAINTENANCE_INTERVAL_SEC)

    def health_probe() -> tuple[bool, dict[str, Any]]:
        try:
            client.ping()
            redis_ok = True
        except Exception:  # noqa: BLE001
            redis_ok = False
        return redis_ok, {
            "worker": config.worker_name,
            "workerId": config.worker_id,
            "queue": queue,
            "redis": "ready" if redis_ok else "down",
            "state": state["status"],
            "reason": state["reason"],
            "processed": state["processed"],
            "failed": state["failed"],
            "capabilities": capabilities,
            "startupError": startup_error,
        }

    health = HealthServer(config.health_port, health_probe)
    health.start()
    threading.Thread(target=heartbeat_loop, daemon=True).start()
    threading.Thread(target=maintenance_loop, daemon=True).start()

    def shutdown(signum: int, _frame: Any) -> None:
        logger.info("Worker faehrt herunter", extra={"extra": {"signal": signum}})
        stop_event.set()

    signal.signal(signal.SIGTERM, shutdown)
    signal.signal(signal.SIGINT, shutdown)

    logger.info(
        "Worker gestartet",
        extra={"extra": {"queue": queue, "storageMode": config.storage_mode, "capabilities": capabilities}},
    )

    while not stop_event.is_set():
        try:
            job = job_queue.reserve(queue, config.worker_id, config.lock_ttl_sec)
        except Exception:  # noqa: BLE001
            logger.error("Queue nicht erreichbar", exc_info=True)
            stop_event.wait(5)
            continue

        if job is None:
            stop_event.wait(POLL_INTERVAL_SEC)
            continue

        state["current_job"] = job.id
        state["status"] = "busy"
        cancelled = threading.Event()
        progress_holder = {"value": 0.0}
        job_stop = threading.Event()

        lock_deadline = time.time() + config.job_timeout_sec

        def lock_loop() -> None:
            while not job_stop.is_set():
                if time.time() >= lock_deadline:
                    logger.error(
                        "Zeitueberschreitung: Sperre wird nicht mehr verlaengert, der Job wird wiederhergestellt",
                        extra={"extra": {"jobId": job.id, "timeoutSec": config.job_timeout_sec}},
                    )
                    return
                try:
                    job_queue.heartbeat(job.id, config.worker_id, config.lock_ttl_sec, int(progress_holder["value"]))
                except Exception:  # noqa: BLE001
                    pass
                job_stop.wait(max(5.0, config.lock_ttl_sec * 0.4))

        lock_thread = threading.Thread(target=lock_loop, daemon=True)
        lock_thread.start()

        def report(progress: float, message: str | None = None) -> None:
            progress_holder["value"] = progress
            if job.ref_id:
                api.report_progress(job.ref_id, progress, message)

        context = JobContext(
            job=job,
            data=job.data,
            config=config,
            api=api,
            storage=storage,
            logger=logger,
            cancelled=cancelled,
            _report=report,
        )

        deadline = time.time() + config.job_timeout_sec
        timer = threading.Timer(config.job_timeout_sec, cancelled.set)
        timer.daemon = True
        timer.start()

        settled = {"done": False}

        def report_safely(what: str, action) -> None:
            try:
                action()
            except Exception as exc:  # noqa: BLE001
                logger.error(
                    "Ergebnis konnte nicht an das Backend gemeldet werden",
                    extra={"extra": {"jobId": job.id, "report": what, "error": str(exc)}},
                )

        try:
            if job.ref_id:
                report_safely(
                    "started",
                    lambda: api.report_started(job.ref_id, {"worker": config.worker_name, "attempt": job.attempts}),
                )
            logger.info("Job gestartet", extra={"extra": {"jobId": job.id, "name": job.name}})

            result = handler(context) or {}

            job_queue.complete(queue, job.id)
            settled["done"] = True
            if job.ref_id:
                report_safely("completed", lambda: api.report_completed(job.ref_id, result))
            state["processed"] += 1
            state["status"] = "idle"
            state["reason"] = startup_error
            logger.info("Job abgeschlossen", extra={"extra": {"jobId": job.id}})

        except WaitingForGpu as exc:
            if settled["done"]:
                logger.error(
                    "Fehler nach erfolgreichem Abschluss: der Job wird nicht erneut eingereiht",
                    extra={"extra": {"jobId": job.id, "error": str(exc)}},
                )
                timer.cancel()
                job_stop.set()
                state["current_job"] = None
                continue
            job_queue.park(queue, job.id, "WAITING_FOR_GPU", str(exc), exc.retry_in_ms)
            settled["done"] = True
            if job.ref_id:
                report_safely("status", lambda: api.report_status(job.ref_id, "WAITING_FOR_GPU", str(exc)))
            state["status"] = "degraded"
            state["reason"] = str(exc)
            logger.warning("Job wartet auf GPU", extra={"extra": {"jobId": job.id, "reason": str(exc)}})

        except Exception as exc:  # noqa: BLE001
            if settled["done"]:
                logger.error(
                    "Fehler nach erfolgreichem Abschluss: der Job wird nicht erneut eingereiht",
                    extra={"extra": {"jobId": job.id, "error": str(exc)}},
                )
                timer.cancel()
                job_stop.set()
                state["current_job"] = None
                continue
            permanent = isinstance(exc, PermanentJobError)
            message = "Zeitueberschreitung" if cancelled.is_set() and time.time() >= deadline else str(exc)
            backoff = min(600000, 30000 * 2 ** max(0, job.attempts - 1))
            outcome = job_queue.fail(queue, job.id, message, backoff, permanent)
            settled["done"] = True
            if job.ref_id:
                report_safely("failed", lambda: api.report_failed(job.ref_id, message, outcome == "RETRY"))
            state["failed"] += 1
            state["status"] = "idle"
            logger.error("Job fehlgeschlagen", exc_info=True, extra={"extra": {"jobId": job.id, "willRetry": outcome == "RETRY"}})

        finally:
            timer.cancel()
            job_stop.set()
            state["current_job"] = None

    try:
        client.delete(worker_key(config.worker_id))
        client.srem(WORKER_SET, config.worker_id)
    except Exception:  # noqa: BLE001
        pass
    health.stop()
    logger.info("Worker gestoppt")

File: packages/worker-core/src/worker.ts
import { Redis } from 'ioredis';
import { BackendApi } from './api-client.js';
import { loadWorkerConfig, type WorkerConfig } from './config.js';
import { startHealthServer, type HealthState } from './health.js';
import { WorkerRegistry } from './heartbeat.js';
import { createLogger, type Logger } from './logger.js';
import { JobQueue } from './queue.js';
import { Storage } from './storage.js';
import {
  PermanentJobError,
  WaitingForGpuError,
  type JobRecord,
  type JobResult,
  type QueueName,
} from './types.js';

const POLL_INTERVAL_MS = 1000;
const MAINTENANCE_INTERVAL_MS = 15_000;
const LOCK_REFRESH_RATIO = 0.4;

export interface JobContext<TData> {
  job: JobRecord<TData>;
  data: TData;
  logger: Logger;
  api: BackendApi;
  storage: Storage;
  config: WorkerConfig;
  /** Wird ausgeloest bei Timeout oder beim Herunterfahren des Containers. */
  signal: AbortSignal;
  reportProgress(progress: number, message?: string): Promise<void>;
}

export type JobHandler<TData> = (ctx: JobContext<TData>) => Promise<JobResult>;

export interface WorkerOptions<TData> {
  /** Anzeigename im Health-Dashboard, z.B. "ffmpeg-worker". */
  name: string;
  queue: QueueName;
  handler: JobHandler<TData>;
  defaultConcurrency?: number;
  /**
   * Einmalige Initialisierung. Der Rueckgabewert landet als "capabilities" im
   * Health-Dashboard (z.B. GPU-Name, ffmpeg-Version).
   */
  onStart?(ctx: { logger: Logger; api: BackendApi; config: WorkerConfig }): Promise<Record<string, unknown>>;
  /** Zusaetzliche Zustandspruefung fuer den Docker-Healthcheck. */
  healthProbe?(): HealthState;
}

export interface RunningWorker {
  stop(): Promise<void>;
}

/**
 * Startet die Worker-Laufzeit: Queue-Polling, Sperr-Heartbeats, Timeouts,
 * Retries, Crash-Recovery, Health-Endpoint und sauberes Herunterfahren.
 *
 * Jeder Worker-Container ruft genau diese Funktion mit seinem Handler auf.
 */
export async function runWorker<TData = Record<string, unknown>>(
  opts: WorkerOptions<TData>,
): Promise<RunningWorker> {
  const config = loadWorkerConfig({ name: opts.name, concurrency: opts.defaultConcurrency });
  const logger = createLogger({ name: config.workerName, workerId: config.workerId, level: config.logLevel });

  const redis = new Redis(config.redisUrl, {
    maxRetriesPerRequest: null,
    enableReadyCheck: true,
    retryStrategy: (times) => Math.min(times * 500, 10_000),
  });
  redis.on('error', (err) => logger.error({ err: err.message }, 'Redis-Fehler'));

  const queue = new JobQueue(redis);
  const api = new BackendApi({
    baseUrl: config.backendUrl,
    apiKey: config.internalApiKey,
    workerId: config.workerId,
    logger,
  });
  const storage = new Storage(config.dataDir, config.storageMode, api);
  const registry = new WorkerRegistry(redis, {
    id: config.workerId,
    name: config.workerName,
    queue: opts.queue,
    version: config.version,
    host: config.host,
    concurrency: config.concurrency,
  });

  let capabilities: Record<string, unknown> = {};
  let startupError: string | null = null;
  if (opts.onStart) {
    try {
      capabilities = await opts.onStart({ logger, api, config });
      registry.setCapabilities(capabilities);
    } catch (err) {
      startupError = (err as Error).message;
      registry.setStatus('degraded', startupError);
      logger.error({ err: startupError }, 'Initialisierung fehlgeschlagen, Worker laeuft eingeschraenkt weiter');
    }
  }

  let shuttingDown = false;
  let active = 0;
  let processed = 0;
  let failed = 0;
  let lastPollOk = Date.now();

  const health = startHealthServer(config.healthPort, () => {
    const custom = opts.healthProbe?.();
    const redisOk = redis.status === 'ready' || redis.status === 'connecting';
    const base: HealthState = {
      ok: redisOk && (custom?.ok ?? true),
      detail: {
        worker: config.workerName,
        workerId: config.workerId,
        queue: opts.queue,
        redis: redis.status,
        active,
        processed,
        failed,
        startupError,
        capabilities,
        lastPollAgoMs: Date.now() - lastPollOk,
        ...(custom?.detail ?? {}),
      },
    };
    return base;
  });

  registry.start();
  logger.info(
    { queue: opts.queue, concurrency: config.concurrency, storageMode: config.storageMode },
    'Worker gestartet',
  );

  async function runJob(job: JobRecord<TData>): Promise<void> {
    active += 1;
    registry.setCurrentJob(job.id);
    const jobLogger = logger.child({ jobId: job.id, jobName: job.name, refId: job.refId });
    const controller = new AbortController();

    const refreshMs = Math.max(5000, config.lockTtlSec * 1000 * LOCK_REFRESH_RATIO);
    let currentProgress = 0;
    const lockTimer = setInterval(() => {
      queue.heartbeat(job.id, config.workerId, config.lockTtlSec, currentProgress).catch(() => undefined);
    }, refreshMs);

    const timeoutTimer = setTimeout(() => controller.abort(), config.jobTimeoutMs);

    const ctx: JobContext<TData> = {
      job,
      data: job.data,
      logger: jobLogger,
      api,
      storage,
      config,
      signal: controller.signal,
      async reportProgress(progress, message) {
        currentProgress = progress;
        await queue.heartbeat(job.id, config.workerId, config.lockTtlSec, progress).catch(() => undefined);
        if (job.refId) await api.reportProgress(job.refId, progress, message);
      },
    };

    let queueSettled = false;

    const report = async (what: string, action: () => Promise<void>): Promise<void> => {
      try {
        await action();
      } catch (err) {
        jobLogger.error(
          { err: (err as Error).message, report: what },
          'Ergebnis konnte nicht an das Backend gemeldet werden',
        );
      }
    };

    try {
      if (job.refId) {
        await report('started', () => api.reportStarted(job.refId!, { worker: config.workerName, attempt: job.attempts }));
      }
      jobLogger.info({ attempt: job.attempts, maxAttempts: job.maxAttempts }, 'Job gestartet');

      const result = await opts.handler(ctx);

      await queue.complete(opts.queue, job.id);
      queueSettled = true;
      if (job.refId) await report('completed', () => api.reportCompleted(job.refId!, result ?? {}));
      processed += 1;
      jobLogger.info('Job erfolgreich abgeschlossen');
    } catch (err) {
      const error = err as Error;

      if (queueSettled) {
        jobLogger.error(
          { err: error.message },
          'Fehler nach erfolgreichem Abschluss: der Job wird nicht erneut eingereiht',
        );
        return;
      }

      if (error instanceof WaitingForGpuError) {
        await queue.park(opts.queue, job.id, 'WAITING_FOR_GPU', error.message, error.retryInMs);
        queueSettled = true;
        if (job.refId) await report('status', () => api.reportStatus(job.refId!, 'WAITING_FOR_GPU', error.message));
        registry.setStatus('degraded', error.message);
        jobLogger.warn({ retryInMs: error.retryInMs }, 'Job wartet auf GPU');
        return;
      }

      const aborted = controller.signal.aborted;
      const message = aborted ? `Zeitueberschreitung nach ${config.jobTimeoutMs} ms` : error.message;
      const permanent = error instanceof PermanentJobError;
      const backoffMs = Math.min(10 * 60_000, 30_000 * 2 ** Math.max(0, job.attempts - 1));

      const outcome = await queue.fail(opts.queue, job.id, message, { backoffMs, permanent });
      queueSettled = true;
      const willRetry = outcome === 'RETRY';
      if (job.refId) await report('failed', () => api.reportFailed(job.refId!, message, willRetry));
      failed += 1;
      jobLogger.error({ err: message, willRetry, permanent }, 'Job fehlgeschlagen');
    } finally {
      clearInterval(lockTimer);
      clearTimeout(timeoutTimer);
      active -= 1;
      registry.setCurrentJob(null);
      if (startupError === null && registry.snapshot().status === 'degraded' && active === 0) {
        registry.setStatus('idle');
      }
    }
  }

  // --- Poll-Schleife --------------------------------------------------------
  const inFlight = new Set<Promise<void>>();

  async function pollOnce(): Promise<void> {
    while (!shuttingDown && inFlight.size < config.concurrency) {
      const job = (await queue.reserve(opts.queue, config.workerId, config.lockTtlSec)) as JobRecord<TData> | null;
      lastPollOk = Date.now();
      if (!job) return;
      const promise = runJob(job)
        .catch((err) => logger.error({ err: (err as Error).message, jobId: job.id }, 'Job-Ausfuehrung abgebrochen'))
        .finally(() => inFlight.delete(promise));
      inFlight.add(promise);
    }
  }

  const pollTimer = setInterval(() => {
    pollOnce().catch((err) => logger.error({ err: (err as Error).message }, 'Poll-Fehler'));
  }, POLL_INTERVAL_MS);

  // Jeder Worker haelt seine eigene Queue instand: faellige Delayed-Jobs
  // aktivieren und Jobs abgestuerzter Worker wiederherstellen.
  const maintenanceTimer = setInterval(() => {
    void (async () => {
      try {
        const promoted = await queue.promoteDelayed(opts.queue);
        const reaped = await queue.reapStalled(opts.queue);
        if (reaped.length > 0) {
          logger.warn({ jobIds: reaped }, 'Jobs abgestuerzter Worker wiederhergestellt');
          await api.log('warn', 'Haengengebliebene Jobs wiederhergestellt', {
            queue: opts.queue,
            jobIds: reaped,
          });
        }
        if (promoted > 0) logger.debug({ promoted }, 'Delayed-Jobs aktiviert');
      } catch (err) {
        logger.debug({ err: (err as Error).message }, 'Wartungslauf fehlgeschlagen');
      }
    })();
  }, MAINTENANCE_INTERVAL_MS);

  async function stop(): Promise<void> {
    if (shuttingDown) return;
    shuttingDown = true;
    logger.info({ inFlight: inFlight.size }, 'Worker faehrt herunter, warte auf laufende Jobs');
    clearInterval(pollTimer);
    clearInterval(maintenanceTimer);
    // Laufende Jobs zu Ende bringen, damit sie nicht als "stalled" gelten.
    await Promise.allSettled([...inFlight]);
    await registry.stop();
    health.close();
    await redis.quit().catch(() => redis.disconnect());
    logger.info('Worker gestoppt');
  }

  for (const signal of ['SIGTERM', 'SIGINT'] as const) {
    process.once(signal, () => {
      void stop().then(() => process.exit(0));
    });
  }

  process.on('unhandledRejection', (reason) => {
    logger.error({ reason: String(reason) }, 'Unbehandelte Promise-Ablehnung im Worker');
  });

  return { stop };
}
